server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/SessionSupplier.scala

```diff
@@ -40,28 +41,38 @@ class SessionSupplier(capabilityFactories: Set[CapabilityFactory], configuration
     .toOption
     .getOrElse(false)
 
-  def generate(username: Username, delegatedUsers: Set[Username], urlPrefixes: UrlPrefixes): Either[IllegalArgumentException, Session] = {
+  def generate(username: Username, delegatedUsers: Set[Username], urlPrefixes: UrlPrefixes): SMono[Session] = {
     val urlEndpointResolver: JmapUrlEndpointResolver = new JmapUrlEndpointResolver(urlPrefixes)
-    val capabilities: Set[Capability] = capabilityFactories
-      .map(cf => cf.create(urlPrefixes, username))
-      .filter(capability => !configuration.disabledCapabilities.contains(capability.identifier()))
 
-    for {
-      account <- accounts(username, capabilities)
-      delegatedAccounts <- delegatedAccounts(delegatedUsers, capabilities)
-    } yield {
-      Session(
-        Capabilities(capabilities),
-        List(account) ++ delegatedAccounts,
-        primaryAccounts(account.accountId, capabilities),
-        username,
-        apiUrl = urlEndpointResolver.apiUrl,
-        downloadUrl = urlEndpointResolver.downloadUrl,
-        uploadUrl = urlEndpointResolver.uploadUrl,
-        eventSourceUrl = urlEndpointResolver.eventSourceUrl)
-    }
+    evaluateCapabilities(username, urlPrefixes)
+      .flatMap { capabilities =>
+        accounts(username, capabilities) match {
+          case Left(e) => SMono.error(e)
+          case Right(account) =>
+            delegatedAccounts(delegatedUsers, capabilities) match {
```

Comment:
two nested match a bit much? Maybe should extract that last bit?

Reply:
Alright, it should look elegant now.